### PR TITLE
Add multiset type and basic casting

### PR DESCRIPTION
**Pull request by yangyangyyy**

### CODE REVIEW

**Comment by electrum** [comment]
"Multiset" is one word, so the "Set" should not be capitalized

File: presto-main/src/main/java/com/facebook/presto/type/MultiSetType.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+
+import java.util.List;
+
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static java.util.Objects.requireNonNull;
+
+public class MultiSetType
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultiSetType.java]: Name this `MultisetType`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToMultiSetCast.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Map;
+
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.Signature.typeParameter;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class ArrayToMultiSetCast
+        extends SqlOperator
+{
+    public static final ArrayToMultiSetCast ARRAY_TO_MULTI_SET_CAST = new ArrayToMultiSetCast();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToMultiSetCast.java]: Name this `ARRAY_TO_MULTISET_CAST`

**Comment by yangyangyyy** [comment]
changed casing to treat multiset as one word

File: presto-main/src/main/java/com/facebook/presto/type/MultisetParametricType.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class MultisetParametricType
+        implements ParametricType
+{
+    public static final MultisetParametricType MULTISET = new MultisetParametricType();
+
+    private MultisetParametricType()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetParametricType.java]: Put the `{}` on the same line:
```java
private MultisetParametricType() {}

File: presto-main/src/main/java/com/facebook/presto/type/MultisetType.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static java.util.Objects.requireNonNull;
+
+public class MultisetType
+        extends AbstractType
+{
+    private final Type elementType;
+    public static final String ARRAY_NULL_ELEMENT_MSG = "ARRAY comparison not supported for arrays with null elements";
+    // use an underlying arraytype for now, to delegate all calls to this.
+    private ArrayType proxy;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: This should be named `delegate`. This class is the proxy.

File: presto-main/src/main/java/com/facebook/presto/type/MultisetType.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.List;
+
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static java.util.Objects.requireNonNull;
+
+public class MultisetType
+        extends AbstractType
+{
+    private final Type elementType;
+    public static final String ARRAY_NULL_ELEMENT_MSG = "ARRAY comparison not supported for arrays with null elements";
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: Don't abbreviate. This should be "message" not "msg"

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: `_ERROR` would be a good choice that matches other places in the code such as in the `Failures` class

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: Actually, this constant appears to be unused

**Comment by cberner** [comment]
@yangyangyyy are you still working on this? If so, could you rebase and then I can take a look

**Comment by yangyangyyy** [comment]
Chris yes I'll rebase .

On Wed, May 4, 2016 at 4:16 PM, Christopher Berner <notifications@github.com

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToMultisetCast.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class ArrayToMultisetCast
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToMultisetCast.java]: Use the new style of function definition, like `ArrayEqualOperator.java`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MultisetToArrayCast.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class MultisetToArrayCast
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MultisetToArrayCast.java]: Use the new style of function definition, like `ArrayEqualOperator.java`

File: presto-main/src/main/java/com/facebook/presto/type/MultisetType.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    // use an underlying arraytype for now, to delegate all calls to this.
+    private ArrayType delegate;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: Lets not do this. Just copy whatever code you need from `ArrayType`

File: presto-main/src/main/java/com/facebook/presto/type/MultisetType.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        delegate = new ArrayType(elementType);
+    }
+
+    public Type getElementType()
+    {
+        return elementType;
+    }
+
+    @Override
+    public boolean isComparable()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: You need to implement `OperatorType.EQUAL` and `OperatorType.HASH_CODE` for any type that is comparable

File: presto-main/src/main/java/com/facebook/presto/type/MultisetType.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public long hash(Block block, int position)
+    {
+        return delegate.hash(block, position);
+    }
+
+    @Override
+    public int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/MultisetType.java]: Remove this since it's not orderable

**Comment by raghavsethi** [comment]
Closing this due to lack of response from author. Please rebase and open another PR you would like this merged.

**Pull request closed without merging**